Font Library modal: Try to improve checkbox labelling

mikachan opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
This PR is an attempt to address https://github.com/WordPress/gutenberg/issues/58299, and covers the following:

- Reverts the changes to the `CheckboxControl` component added in https://github.com/WordPress/gutenberg/pull/56158
- Tries an alternative labelling solution for the checkboxes in the Font Library modal without needing to adjust the `CheckboxControl` component

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
To improve accessibility and usability of these checkboxes whilst maintaining the functionality of the `CheckboxControl` component.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
The main challenge with the current implementation of the Font Library checkboxes is that the visible label can either be text or an image, depending on if the font face provides a preview image (e.g. an SVG). This logic cannot currently be passed to the `CheckboxControl` component, as the `label` prop only accepts a string (which I agree with).

Also, the older design for the font library checkboxes had the checkboxes to the right of the label, rather than the left. Now the design has been updated to be in line with the `CheckboxControl` component, where the checkbox is to the left of the label. This makes it much easier to use the existing functionality of this component.

The solution in this PR is to provide a label in the `CheckboxControl`, and then hide the label element that the component creates using CSS. The CSS should visibly hide the label while still allowing it to be accessed by screen-readers. This has worked well in my testing so far.

There has also been an `onClick` handler added to the `FontFaceDemo` component, which allows the checkbox to be toggled when this component is clicked. Otherwise, only the checkbox itself can be interacted with to toggle the checkbox value, and I don't think this is good for accessibility or usability.

This is one attempt at an alternative solution to handling these checkbox labels, and I'd love feedback on whether this is a good approach or if we should try alternatives.

Possible other alternatives:
- Use plain text as the label, and move the font demo to a separate element (e.g. a column layout, where the font demo sits to the right of the text label). This would require design input.
- Add `aria-labelledby` attribute to the `CheckboxControl`, and reference the `FontFaceDemo` element as the label
- Allow a custom label functionality to be passed to the `CheckboxControl` component (that doesn't allow the label to be `false`)

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->
1. Ensure you're using a theme that includes at least one bundled font family (or upload a font family as part of this test)
2. Open the Font Library modal
3. Open a font family
4. Toggle a font variant of the selected font family by clicking either the checkbox or the text/image font preview

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->
1. Ensure you're using a theme that includes at least one bundled font family (or upload a font family as part of this test)
2. Open the Font Library modal
3. Open a font family
4. Toggle a font variant of the selected font family by tabbing to the font variant panel and toggling the checkbox. If using a screenreader, the font variant name should be called out.

## Screenshots or screencast <!-- if applicable -->

https://github.com/WordPress/gutenberg/assets/1645628/114bb980-1232-4933-9bdb-9ab84ef733b3

mikachan committed:
Revert allowing label to be false

mikachan committed:
Replace custom labels with label prop

mikachan committed:
Try using aria-labelledby

mikachan committed:
Make FontDemo clickable

mikachan committed:
Try hiding label with CSS

mikachan committed:
Update components changelog

mikachan committed:
Fix font library e2e test

mikachan committed:
Remove checkboxId prop

mirka reviewed (commented):
Thanks for the follow-up! The revert to the CheckboxControl API changes look good.

On the Font Library side of things, the approach feels a little bit roundabout to me, especially with the faux button we're trying to do in `FontFaceDemo`. I just proposed adding a custom label example to the Storybook in #58438 that I think is the most straightforward approach in most cases, similar to one of the alternative methods you suggested in the PR description. Using a `<label>` element would avoid having to implement a faux button, and since an `<img>` is phrasing content, I think it should be permitted to include inside a label node.

What do you think?

mikachan commented:
Thanks for taking a look, @mirka!

> the approach feels a little bit roundabout to me, especially with the faux button we're trying to do in FontFaceDemo

I agree, although I tried the quickest/most simple solution first 😅 This is also why I attempted my original approach in https://github.com/WordPress/gutenberg/issues/58299, as I'd rather take advantage of all functionality already available using a native `label`.

Thanks so much for the example in https://github.com/WordPress/gutenberg/pull/58438! Although, does this not render two `label`s, one empty tag from the `CheckboxControl` component and then the second, custom `label`?

mirka commented:
> Although, does this not render two `label`s, one empty tag from the `CheckboxControl` component and then the second, custom `label`?

Good point, I was still referring to the implementation in trunk. I think the [conditional rendering logic](https://github.com/WordPress/gutenberg/commit/d9033b7ad18fb73160d8018fb8f6da9b81419841#diff-ad327c4ba4df23e9aa8712664f69691187c90f18c54521630d421040dd24ffa5) you added in #56158 was great — that change was basically a bug fix, since rendering an empty label is nonsensical. It's just the type changes that were unnecessary (i.e. adding a `false` type to the `label` prop). So I suggest we don't revert the [conditional](https://github.com/WordPress/gutenberg/pull/58339/files#diff-ad327c4ba4df23e9aa8712664f69691187c90f18c54521630d421040dd24ffa5).

ciampo commented:
One of the main pieces of feedback received on https://github.com/WordPress/gutenberg/pull/56158 was that we were allowing consumers of `CheckboxControl` to disable the `label` entirely by passing `false`.

With the latest suggested changes, users would still be able to do so by omitting the `label` prop. I know that this isn't much different (it's probably worse!) than having an empty label, but I wonder if we can do better to make sure that consumers of `CheckboxControl` provide a label for the control.

A few ideas that came to mind:

- have a runtime check that checks if there's a `label` with the `for` attribute targeting the checkbox `input`, and outputs a warning if it doesn't find one
- allow consumers to pass a classname to the label, so that it can styled with CSS
- change the type of label so that it accepts a `ReactNode` (or something similar). Consumers who need more flexibility would be able to pass a new `label` and would therefore have more freedom in styling it

In any case, we should make sure that consumers of `CheckboxControl` are able to specify the `id` of the underlying `input` element

mirka commented:
For the scope of this PR, I suggest limiting the CheckboxControl changes to reverting the addition of a `false` type, which was simply unnecessary.

Anything else is enhancement territory, which should be discussed separately IMO 🙂

mikachan committed:
Conditionally render label

mikachan committed:
Try using separate label tag

mikachan committed:
Update e2e test

mikachan commented:
Thanks so much, both!

I agree with the idea not to increase the scope of this PR too much, so I've gone ahead and added the conditional rendering of the label tag in `CheckboxControl` in this commit: https://github.com/WordPress/gutenberg/pull/58339/commits/d53293a702912ac41e6c78315f3bc5247bea6512 (and the `false` type is still being removed.)

And then I've updated the Font Library checkbox markup in line with your example above, so there is now a separate label tag that holds the preview image or text. Committed here: https://github.com/WordPress/gutenberg/pull/58339/commits/239d4f4b629135672419db208d85594d56f1446f

This tests well for me and I feel much happier with this approach. Thank you for helping me work through it 🙇

mikachan committed:
Update snapshot

mikachan committed:
Update CheckboxControl test

ciampo reviewed (approved):
Code changes LGTM and follow @mirka 's suggestions above 🚀 

Thank you for working on this 🙏
